### ISSUE TITLE
Working with boreholes

### ISSUE THREAD

**Issue by hntigkakis** [cgre-aachen/gempy#669]
Hello! I came across Gempy recently and I want to use it to create a model of both bedrock and superficial deposits. My data consist of a large number of borehole records that I am using to extract the surface points. However, I am having a hard time finding the correct overall order of the geology formations and seeing which overlies which. Is there a way to look at all the borehole records and extract the correct order of the formations?

**Comment by Japhiolite** [comment]
Hi @hntigkakis 

a dedicated borehole integration to gempy can be achieved via the libraries [striplog](https://github.com/agile-geoscience/striplog), vor instance. Here's an example of how [striplog can be combined with gempy](https://docs.gempy.org/integrations/gempy_striplog.html). So essentially, you need to put some preprocessing work into cleaning the data. Striplog is an excellent library for that.

Best,
Jan

**Comment by hntigkakis** [comment]
Hi @Japhiolite 

Thank you for your reply. My problem has more to do with determining the stratigraphy, after the data has been cleaned up. See bellow for an example:

![Screenshot 2022-02-21 121417](https://user-images.githubusercontent.com/99178328/154953355-7a3a6120-869d-4386-8518-375bd8389d0f.png)

Simply by visually examining these two boreholes, I would create a gempy model that would look like this:

![Untitled](https://user-images.githubusercontent.com/99178328/154955321-6ec08108-a4be-43f2-8c9f-e19cfdf15f7f.png)

My problem however is that in this case, I did this process manually, by examining just two boreholes. Had I just imported these two logs into gempy, it would have created a single CLAY surface, when in this case there are 3. Considering that I have over 100 boreholes, is there a way to automate this? Is there a way to tell the script that if there is another borehole with a CLAY surface point, that point belongs to CLAY-1, CLAY-2, CLAY-3, or in a completely new surface called CLAY-4?

**Comment by Japhiolite** [comment]
Hi @hntigkakis 

There certainly is a way of doing this, but I view this more as a thing I'd do **before** I load the interfaces in GemPy, as a sort of preprocessing. Thus I'd approach this using pandas (and possibly further other libraries).

After a quick google search, the top most response by anky in this stack overflow post looks like it can be useful to you:
https://stackoverflow.com/questions/59951415/how-do-you-replace-duplicate-values-with-multiple-unique-strings-in-pandas

There the question is a similar one to yours.

**Comment by Japhiolite** [comment]
Hi @hntigkakis just to catch up: Did that suggestion help you in creating the input data to your needs (with subscripts to repeating lithologies)?

**Comment by hntigkakis** [comment]
Hi @Japhiolite. 

Yes and no. It solves the problem of sorting the surface names, but I still can't figure out how to determine the order of the formations. From my understanding, Gempy needs an "overall stratigraphy" or "overall order" (which is what the surfaces list in the model is). CLAY-2 and COAL might not cover the same parts of the model domain, but Gempy needs to know that where there are present, CLAY-2 is higher than COAL.

In my example, I created this "overall stratigraphy" by matching patterns between the two boreholes. Going from top to bottom, both boreholes follow the same pattern and have the same formations in the same order. Then I meet COAL and SILTSTONE-2 in borehole A but not in borehole B. And then the pattern continues again. By doing that, I determined that COAL and SILTSTONE-2 were between SILTSTONE and SANDSTONE-2.

Doing that in this example is easy because there are only two boreholes with only two formations "out of place". Now if we add more boreholes into the analysis (in my case I have around 100), is where we come across my problem. How do you determine this ordered list of the "overall stratigraphy"? 

One solution is to simplify the problem by grouping geological formations together, or by ignoring formations with very small thickness. While that approach simplifies the problem by reducing the overall number of surfaces, it still doesn't solve the problem of putting those surfaces in the correct order.

**Comment by Japhiolite** [comment]
Ah, ok I see. 

I think the solution you propose at the end is in fact a question, that should be asked at the beginning: What's the purpose of the model you want to create?
Depending on the purpose, you'd have to standardize your boreholes in a different fashion. Would the purpose be a general geological model? Then uniting lithologies into the same geological formation would make sense. Is the purpose translating the model to a geometry for hydrogeological simulation? Then lithologies should be combined depending on their petrophysical parameters (i.e. defining a hydrostratigraphy).

GemPy itself doesn't have a geologist's point of view, so cannot tell if a surface stack ( a model stratigraphy) is reasonable or not.
Standardizing your borehole stratigraphy to a defined model stratigraphy is a step, I think is difficult to speed up automatically. Striplogs use different color coding or hashes to mark the same lithology across different borehole. 

You could implement something with does not only look at the current lithology, but at a stack of, let's say 3 to 5 lithologies, and tries to identify the similarity between different stacks to come to a conclusion if, e.g. SANDSTONE-2 in well1 is correlated to SANDSTONE-3 in well2.
But due to the problems you described, i.e. the spatial heterogeneity and varying extent of geological units, there's still a good portion of manual work involved, I fear. 

Summing up: Creating an overall model stratigraphy and changing the borehole data you have towards that model stratigraphy is a necessary preprocessing step, which can be quite labor intensive.

**Comment by hntigkakis** [comment]
Hi @Japhiolite 

I am still struggling to be honest. As you pointed out, I need a suitable (and efficient) way to pre-process my data that I haven't figured out. The end goal is to develop a groundwater model, so the ultimate question is the amount of detail I need my geology model to have to fit the purpose of my research.

In any case, I'll close this question as it's not about GemPy anymore, but modelling in general. Thank you for all your help and I'll be back when I get stuck again
